include/dali/c_api.h

```diff
@@ -23,214 +23,239 @@
-   */
-  DLL_PUBLIC void daliDeletePipeline(daliPipelineHandle* pipe_handle);
-
-  /**
-   * @brief Load plugin library
-   */
-  DLL_PUBLIC void daliLoadLibrary(const char* lib_path);
+
+/**
+ * @brief Handle for DALI C-like API.
+ *
+ * @note Beware, the C API is just C-like API for handling some mangling issues and
+ * it can throw exceptions.
+ */
+typedef struct {
+  void *pipe;
+  void *ws;
+} daliPipelineHandle;
+
+typedef enum {
+  CPU = 0,
+  GPU = 1
+} device_type_t;
+
+typedef enum {
+  DALI_NO_TYPE = -1,
+  DALI_UINT8 = 0,
+  DALI_UINT16 = 1,
+  DALI_UINT32 = 2,
+  DALI_UINT64 = 3,
+  DALI_INT8 = 4,
+  DALI_INT16 = 5,
+  DALI_INT32 = 6,
+  DALI_INT64 = 7,
+  DALI_FLOAT16 = 8,
+  DALI_FLOAT = 9,
+  DALI_FLOAT64 = 10,
+  DALI_BOOL = 11
+} dali_data_type_t;
+
+/**
+ * @brief Create DALI pipeline. Setting batch_size,
+ * num_threads or device_id here overrides
+ * values stored in the serialized pipeline.
+ * When separated_execution is false, prefetch_queue_depth is considered,
+ * gpu_prefetch_queue_depth and cpu_prefetch_queue_depth are ignored.
+ * When separated_execution is true, cpu_prefetch_queue_depth and
+ * gpu_prefetch_queue_depth are considered and prefetch_queue_depth is ignored.
+ */
+DLL_PUBLIC void daliCreatePipeline(daliPipelineHandle *pipe_handle,
+                                   const char *serialized_pipeline,
+                                   int length,
+                                   int batch_size,
+                                   int num_threads,
+                                   int device_id,
+                                   bool separated_execution,
+                                   int prefetch_queue_depth,
+                                   int cpu_prefetch_queue_depth,
+                                   int gpu_prefetch_queue_depth);
+
+/**
+ * @brief Feed the data to ExternalSource as contiguous memory.
+ *
+ * When calling this function for GPU device, default stream (0) will be used when copying data
+ * onto GPU. An alternative function exists to specify also a CUDA stream.
+ * If used with CPU device, stream value will be ignored
+ *
+ * @param pipe_handle Pointer to pipeline handle
+ * @param name Pointer to a null-terminated byte string with the name of the External Source
+ *             to be fed
+ * @param device Device of the supplied memory. Only CPU is supported.
+ * @param data_ptr Pointer to contiguous buffer containing all samples
+ * @param data_type Type of the provided data
+ * @param shapes Pointer to an array containing shapes of all samples concatenated one after
+ *              another. Should contain batch_size * sample_dim elements.
+ * @param sample_dim The dimensionality of a single sample.
+ * @param layout_str Optional layout provided as a pointer to null-terminated byte string.
+ *                   Can be set to NULL.
+ * @param stream CUDA stream to use when copying the data onto GPU. If used with `device == CPU`,
```

Comment:
There is no such arg in this function. Still you should state that for the GPU device it will use a default stream. If the user wants different one `daliSetExternalInputStream` should be used.

Reply:
Introduced doxygen's member groups (`///@{`). Also, I changed the name to `daliSetExternalInputCudaStream`, cause with the previous one it sounded like the `input stream` would be set using this function.